### ISSUE TITLE
clang crash when compiling to wasm32 with multi-value

### ISSUE THREAD

**Issue by TerrorJack** [llvm/llvm-project#58297]
Minimal repro C source:

```c
struct a {
  int b;
};
int c;
double d;
long double e() {
  struct a f;
  if (d)
    return 0;
  if (c)
    f.b = d;
  return f.b;
}
```

Minimal repro command-line:

```sh
/usr/lib/llvm-15/bin/clang --target=wasm32-wasi -O1 -mmultivalue -c /tmp/nextafterl.c
```

Error message:

```console
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /usr/lib/llvm-15/bin/clang --target=wasm32-wasi -O1 -mmultivalue -c /tmp/nextafterl.c
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module '/tmp/nextafterl.c'.
4.      Running pass 'WebAssembly Register Coloring' on function '@e'
 #0 0x00007f37c8b1a451 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xf04451)
 #1 0x00007f37c8b1818e llvm::sys::RunSignalHandlers() (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xf0218e)
 #2 0x00007f37c8b19801 llvm::sys::CleanupOnSignal(unsigned long) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xf03801)
 #3 0x00007f37c8a3d70f (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xe2770f)
 #4 0x00007f37c76ef520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #5 0x00007f37cb4287df (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x38127df)
 #6 0x00007f37cb4274cd (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x38114cd)
 #7 0x00007f37c8eaa00c llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x129400c)
 #8 0x00007f37c8c56d72 llvm::FPPassManager::runOnFunction(llvm::Function&) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x1040d72)
 #9 0x00007f37c8c5e333 llvm::FPPassManager::runOnModule(llvm::Module&) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x1048333)
#10 0x00007f37c8c57916 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x1041916)
#11 0x00007f37d05298b5 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>) (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x19228b5)
#12 0x00007f37d0889701 (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x1c82701)
#13 0x00007f37cf693ebb clang::ParseAST(clang::Sema&, bool, bool) (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0xa8cebb)
#14 0x00007f37d0885775 clang::CodeGenAction::ExecuteAction() (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x1c7e775)
#15 0x00007f37d12bb0a7 clang::FrontendAction::Execute() (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x26b40a7)
#16 0x00007f37d122e1b6 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x26271b6)
#17 0x00007f37d133904a clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x273204a)
#18 0x000055c44e1939e0 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/lib/llvm-15/bin/clang+0x139e0)
#19 0x000055c44e191bab (/usr/lib/llvm-15/bin/clang+0x11bab)
#20 0x00007f37d0e929b2 (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x228b9b2)
#21 0x00007f37c8a3d46e llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xe2746e)
#22 0x00007f37d0e92478 clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x228b478)
#23 0x00007f37d0e5486a clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x224d86a)
#24 0x00007f37d0e54abe clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x224dabe)
#25 0x00007f37d0e718ff clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x226a8ff)
#26 0x000055c44e191311 clang_main(int, char**) (/usr/lib/llvm-15/bin/clang+0x11311)
#27 0x00007f37c76d6d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#28 0x00007f37c76d6e40 call_init ./csu/../csu/libc-start.c:128:20
#29 0x00007f37c76d6e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#30 0x000055c44e18e3b5 _start (/usr/lib/llvm-15/bin/clang+0xe3b5)
clang: error: clang frontend command failed with exit code 139 (use -v to see invocation)
Ubuntu clang version 15.0.2-++20221003113220+5f620d003fda-1~exp1~20221003113231.72
Target: wasm32-unknown-wasi
Thread model: posix
InstalledDir: /usr/lib/llvm-15/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/tmp.tTiXXhik2w/nextafterl-3d9fcd.c
clang: note: diagnostic msg: /tmp/tmp.tTiXXhik2w/nextafterl-3d9fcd.sh
clang: note: diagnostic msg: 

********************
```

**Comment by sparker-arm** [comment]
Proposed fix: https://reviews.llvm.org/D137824

**Comment by sparker-arm** [comment]
This also looks like a duplicate of this [issue](https://github.com/llvm/llvm-project/issues/55136).

**Comment by sparker-arm** [comment]
Fixed in commit [b303c0027ff7f3a9912d7690886b7b7b33ddb05f](https://github.com/llvm/llvm-project/commit/b303c0027ff7f3a9912d7690886b7b7b33ddb05f)
